parquet/src/main/java/org/apache/iceberg/parquet/ParquetTypeVisitor.java

```diff
@@ -123,19 +60,115 @@
     }
   }
 
-  private static <T> T visitField(Type field, ParquetTypeVisitor<T> visitor) {
-    visitor.fieldNames.push(field.getName());
+  private static <T> T visitList(GroupType list, ParquetTypeVisitor<T> visitor) {
+    Preconditions.checkArgument(!list.isRepetition(Type.Repetition.REPEATED),
+        "Invalid list: top-level group is repeated: %s", list);
+    Preconditions.checkArgument(list.getFieldCount() == 1,
+        "Invalid list: does not contain single repeated field: %s", list);
+
+    GroupType repeatedElement = list.getFields().get(0).asGroupType();
+    Preconditions.checkArgument(repeatedElement.isRepetition(Type.Repetition.REPEATED),
+        "Invalid list: inner group is not repeated");
+    Preconditions.checkArgument(repeatedElement.getFieldCount() <= 1,
+        "Invalid list: repeated group is not a single field: %s", list);
+
+    visitor.beforeRepeatedElement(repeatedElement);
     try {
-      return visit(field, visitor);
+      T elementResult = null;
+      if (repeatedElement.getFieldCount() > 0) {
+        Type elementField = repeatedElement.getType(0);
+        visitor.beforeElementField(elementField);
+        try {
+          elementResult = visit(elementField, visitor);
+        } finally {
+          visitor.afterElementField(elementField);
+        }
+      }
+
+      return visitor.list(list, elementResult);
+
     } finally {
-      visitor.fieldNames.pop();
+      visitor.afterRepeatedElement(repeatedElement);
+    }
+  }
+
+  private static <T> T visitMap(GroupType map, ParquetTypeVisitor<T> visitor) {
```

Comment:
This was moved to its own method. The only changes are to use the new before and after methods.